clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp

```diff
@@ -790,6 +773,65 @@ bool clang::dpct::BarrierFenceSpaceAnalyzer::isAccessingMemory(
   return false;
 }
 
+bool clang::dpct::BarrierFenceSpaceAnalyzer::isInRanges(
+    SourceLocation SL, std::vector<SourceRange> Ranges) {
+  auto &SM = DpctGlobalInfo::getSourceManager();
+  for (auto &Range : Ranges) {
+    if (SM.getFileOffset(Range.getBegin()) < SM.getFileOffset(SL) &&
+        SM.getFileOffset(SL) < SM.getFileOffset(Range.getEnd())) {
+      return true;
+    }
+  }
+  return false;
+}
+
+std::pair<bool, std::string>
+clang::dpct::BarrierFenceSpaceAnalyzer::isSafeWriteAfterWrite(
+    const std::set<const DeclRefExpr *> &WAWDRESet) {
+  if (WAWDRESet.size() > 1) {
+#ifdef __DEBUG_BARRIER_FENCE_SPACE_ANALYZER
+    std::cout << "isSafeWriteAfterWrite False case 1" << std::endl;
+#endif
+    return {false, ""};
+  }
+
+  if (!IsDifferenceBetweenThreadIdxXAndIndexConstant) {
+#ifdef __DEBUG_BARRIER_FENCE_SPACE_ANALYZER
+    std::cout << "isSafeWriteAfterWrite False case 2" << std::endl;
+#endif
+    return {false, ""};
+  }
+  const DeclRefExpr *DRE = *WAWDRESet.begin();
+
+  // If only in single loop
+  auto FirstLoop = findNearestLoopStmt(DRE);
+  if (!FirstLoop) {
+#ifdef __DEBUG_BARRIER_FENCE_SPACE_ANALYZER
+    std::cout << "isSafeWriteAfterWrite False case 3" << std::endl;
+#endif
+    return {false, ""};
+  }
+  auto SecondLoop = findNearestLoopStmt(FirstLoop);
+  if (SecondLoop) {
+#ifdef __DEBUG_BARRIER_FENCE_SPACE_ANALYZER
+    std::cout << "isSafeWriteAfterWrite False case 4" << std::endl;
+#endif
+    return {false, ""};
+  }
+
+  // Find step
+  // FIXME: This DRE is "next" in next[idx], we need find DRE of idx in "idx +=
+  // nx"
+  auto Res = getIdxOfASEConstValueExpr(getArraySubscriptExpr(DRE));
```

Comment:
Both `getIdxOfASEConstValueExpr` and `getArraySubscriptExpr` have been already called before, can we save the result at that time to avoid duplication?